### PR TITLE
[WIP] Add a default sliding expiration of 30 seconds on Memory and Distributed Cache Tag Helpers.

### PR DESCRIPTION
**Pull request by javiercn**
Adds a default sliding expiration for the cache tag helpers and creates a private memorycache instance to use with CacheTagHelper

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/DependencyInjection/TagHelperExtensions.cs [hunk 1]
```diff
@@ -32,7 +35,10 @@ public static IMvcCoreBuilder AddCacheTagHelper(this IMvcCoreBuilder builder)
 
             // Required default services for cache tag helpers
             builder.Services.TryAddSingleton<IDistributedCache, MemoryDistributedCache>();
-            builder.Services.TryAddSingleton<IMemoryCache, MemoryCache>();
+            builder.Services.TryAddSingleton<CacheTagHelperMemoryCacheFactory>();
+            //builder.Services.Configure<MemoryCacheOptions>(
+            //    CacheTagHelperMemoryCacheFactory.MemoryCacheOptionsName,
+            //    o => o.EntryCountLimit = DefaultEntryCountLimit);
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/DependencyInjection/TagHelperExtensions.cs]: Waiting for the changes in Caching

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/Internal/CacheTagHelperMemoryCacheFactory.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            Cache = new MemoryCache(optionsMonitor.Get(MemoryCacheOptionsName));
+        }
+
+        // For testing only.
+        internal CacheTagHelperMemoryCacheFactory(IMemoryCache cache)
+        {
+            Cache = cache;
+        }
+
+        public IMemoryCache Cache { get; }
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/Internal/CacheTagHelperMemoryCacheFactory.cs]: Should we make this get;set to let people provide their own IMemoryCache implementation (not sure how common a scenario that would be, but I wouldn't think it's common).
Alternatively, just make the internal constructor public to allow people to register this type in DI with their own IMemoryCache instance.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/Internal/CacheTagHelperMemoryCacheFactory.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers.Internal
+{
+    public class CacheTagHelperMemoryCacheFactory
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/Internal/CacheTagHelperMemoryCacheFactory.cs]: Let me know if you want unit tests for this. I believe it's ok as it's covered by functional tests.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs [hunk 4]
```diff
@@ -147,27 +149,37 @@ public CacheTagHelper(IMemoryCache memoryCache, HtmlEncoder htmlEncoder) : base(
...
                 options.SetSlidingExpiration(ExpiresSliding.Value);
             }
 
             if (Priority != null)
             {
+                hasEvictionCriteria = true;
                 options.SetPriority(Priority.Value);
             }
 
+            if (!hasEvictionCriteria)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs]: Should this be baked into MemoryCache? Something along the lines of 
```
DefaultEvictionPolicy
{
   Absolute, Sliding, None
}
```

and consumers opt out if they explicitly want long-lived caches?

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs [hunk 5]
```diff
@@ -147,27 +149,37 @@ public CacheTagHelper(IMemoryCache memoryCache, HtmlEncoder htmlEncoder) : base(
...
 
             if (ExpiresSliding != null)
             {
+                hasEvictionCriteria = true;
                 options.SetSlidingExpiration(ExpiresSliding.Value);
             }
 
             if (Priority != null)
             {
+                hasEvictionCriteria = true;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs]: Don't think this is an eviction criteria.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs]: I don't think `Priority` should be considered an eviction criteria. Looking at https://docs.microsoft.com/en-us/aspnet/core/api/microsoft.extensions.caching.memory.memorycacheentryoptions

It doesn't specify when something should be evicted, only its relative importance.

Also we removed the memory-pressure-based automatic eviction, so this won't happen without someone triggering it.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelperBase.cs [hunk 6]
```diff
@@ -99,6 +99,14 @@ public CacheTagHelperBase(HtmlEncoder htmlEncoder)
         public TimeSpan? ExpiresAfter { get; set; }
 
         /// <summary>
+        /// Gets or sets the default duration, from the time the cache entry was added, when it should be evicted.
+        /// This default duration will only be used if no other expiration criteria gets set for the tag helper.
+        /// The default expiration time is a sliding expiration of 30 seconds.
+        /// </summary>
+        [HtmlAttributeNotBound]
+        public TimeSpan DefaultExpiration { get; set; } = TimeSpan.FromSeconds(30);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelperBase.cs]: suggest making this public static readonly field - there's no real scenario for changing it, you would set one of the other properties instead

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/DependencyInjection/TagHelperExtensions.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/DependencyInjection/TagHelperExtensions.cs]: Let's go ahead with this PR without waiting for `EntryCountLimit` to be added. I want to get as much of this in today as possible.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/Internal/CacheTagHelperMemoryCacheFactory.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/Internal/CacheTagHelperMemoryCacheFactory.cs]: > Alternatively, just make the internal constructor public to allow people to register this type in DI with their own IMemoryCache instance.

Making this public means we have two 'DI' constructors. What about a static factory method? `CTHMCFactory.Create(IMemoryCache ...)`

Or you can make `CacheTagHelperMemoryCacheFactory(IMemoryCache cache)` public, delete `CacheTagHelperMemoryCacheFactory(IOptionsMonitor<MemoryCacheOptions> optionsMonitor)` and register this in DI using a factory method.

**Comment by javiercn** [comment]
Updated

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs (hunk 4 above)
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs]: I don't think you can have a generic eviction policy for the memory cache that is effective. That would be a separate and orthogonal thing, but IMO the policy should be set by the component that makes use of the cache, as it is that component the one that knows about the parameters of the things that its saving on it.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/DistributedCacheTagHelperTest.cs [hunk 7]
```diff
@@ -272,6 +272,29 @@ public void UpdateCacheEntryOptions_SetsAbsoluteExpiration_IfExpiresOnIsSet()
         }
 
         [Fact]
+        public void UpdateCacheEntryOptions_DefaultsTo30SecondsSliding_IfNoEvictionCriteriaIsProvided()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/DistributedCacheTagHelperTest.cs]: Can you add a test where the priority is specified (or just amend this to include it)/

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelperBase.cs [hunk 8]
```diff
@@ -26,6 +26,13 @@ public abstract class CacheTagHelperBase : TagHelper
         private const string EnabledAttributeName = "enabled";
 
         /// <summary>
+        /// Gets or sets the default duration, from the time the cache entry was added, when it should be evicted.
+        /// This default duration will only be used if no other expiration criteria gets set for the tag helper.
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelperBase.cs]: > is no other expiration criteria ~~gets set for the tag helper~~ is specified

Up to you if you want to change it. Not picky 👍

**Pull request merged**